File: push/src/main/java/io/appmetrica/analytics/push/impl/PushServiceControllerComposite.java
package io.appmetrica.analytics.push.impl;

import android.content.Context;
import androidx.annotation.NonNull;
import io.appmetrica.analytics.push.coreutils.internal.PushServiceFacade;
import io.appmetrica.analytics.push.provider.api.PushServiceController;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class PushServiceControllerComposite {

    @NonNull
    private final Context context;
    @NonNull
    private final Collection<PushServiceController> controllers;
    @NonNull
    private final Collection<String> titles;

    PushServiceControllerComposite(@NonNull Context context, @NonNull Collection<PushServiceController> controllers) {
        this.context = context;
        this.controllers = Collections.unmodifiableList(new ArrayList<PushServiceController>(controllers));
        ArrayList<String> tmpTitles = new ArrayList<String>(controllers.size());
        for (PushServiceController controller : controllers) {
            tmpTitles.add(controller.getTitle());
        }
        titles = Collections.unmodifiableList(tmpTitles);
    }

    public void register() {
        boolean wasRegistered = false;
        for (PushServiceController controller : controllers) {
            wasRegistered |= controller.register();
        }
        if (wasRegistered) {
            PushServiceFacade.initToken(context);
        }
    }

    @NonNull
    public Map<String, String> getTokens() {
        HashMap<String, String> tokens = new HashMap<String, String>();
        for (PushServiceController controller : controllers) {
            String title = controller.getTitle();
            String token = controller.getToken();
            tokens.put(title, token);
        }
        return Collections.unmodifiableMap(tokens);
    }

    @NonNull
    public Collection<String> getTitles() {
        return titles;
    }

}

File: push/src/main/java/io/appmetrica/analytics/push/impl/PushServiceControllerComposite.kt
package io.appmetrica.analytics.push.impl

import android.content.Context
import io.appmetrica.analytics.push.coreutils.internal.PushServiceFacade
import io.appmetrica.analytics.push.provider.api.PushServiceController
import io.appmetrica.analytics.push.provider.api.PushServiceExecutionRestrictions

internal class PushServiceControllerComposite internal constructor(
    private val context: Context,
    controllers: List<PushServiceController>
) {

    private val controllers: Map<String, PushServiceController> = controllers.associateBy { it.transportId }
    val transportIds: Set<String> get() = this.controllers.keys

    val tokens: Map<String, String?> get() = this.controllers.mapValues { it.value.token }

    fun register() {
        controllers.values.forEach { controller ->
            if (controller.register()) {
                PushServiceFacade.initToken(context, controller.transportId)
            }
        }
    }

    fun getToken(provider: String): String? = controllers[provider]?.token

    fun getExecutionRestrictions(transport: String): PushServiceExecutionRestrictions? =
        controllers[transport]?.executionRestrictions

    fun shouldSendTokenForProvider(token: String, provider: String) =
        controllers[provider]?.shouldSendToken(token) ?: false
}
